src/main/java/com/shapesecurity/salvation/data/Policy.java

```diff
@@ -195,20 +190,70 @@ private void optimise() {
             }
 
             // * remove unnecessary default-src directives if all source directives exist
-            if (this.directives.containsKey(ScriptSrcDirective.class) &&
-                this.directives.containsKey(StyleSrcDirective.class) &&
-                this.directives.containsKey(ImgSrcDirective.class) &&
-                this.directives.containsKey(ChildSrcDirective.class) &&
-                this.directives.containsKey(ConnectSrcDirective.class) &&
-                this.directives.containsKey(FontSrcDirective.class) &&
-                this.directives.containsKey(MediaSrcDirective.class) &&
-                this.directives.containsKey(ObjectSrcDirective.class) &&
-                this.directives.containsKey(ManifestSrcDirective.class)) {
+            if (fold(Directive.getFetchDirectives(), (memo, fetchDirective) -> memo && this.directives.containsKey(fetchDirective), true)) {
```

Comment:
Instead of using `fold` here, I think we should just make a function `<A> boolean all(List<A> list, Function<A, Boolean> predicate);` so we can short-circuit.